kj-rs/convert.h

```diff
@@ -0,0 +1,227 @@
+  template <typename T>
+  static ::rust::Slice<const T> from(const kj::ArrayPtr<T>* arr) {
+    return ::rust::Slice<const T>(arr->begin(), arr->size());
+  }
+
+  /// kjArray.as<Rust>() - via Rust::from(&kjArray)
+  template <typename T>
+  static ::rust::Slice<const T> from(const kj::Array<T>* arr) {
+    return ::rust::Slice<const T>(arr->begin(), arr->size());
+  }
+
+  /// kjString.as<Rust>() - via Rust::from(&kjString)
+  static ::rust::String from(const kj::String* str) {
+    return ::rust::String(str->begin(), str->size());
+  }
+
+  /// kjStringPtr.as<Rust>() - via Rust::from(&kjStringPtr)
+  static ::rust::Str from(const kj::StringPtr* str) {
+    return ::rust::Str(str->begin(), str->size());
+  }
+
+  // into() methods for from<Rust>(rustObject) - converting Rust to KJ
+
+  /// from<Rust>(rustVec) - Zero-copy read-only view
+  template <typename T>
+  static kj::ArrayPtr<const T> into(const ::rust::Vec<T>& vec) {
+    return kj::ArrayPtr<const T>(vec.data(), vec.size());
+  }
+
+  /// from<Rust>(rustSlice) - Zero-copy slice view
+  template <typename T>
+  static kj::ArrayPtr<T> into(const ::rust::Slice<T>& slice) {
+    return kj::ArrayPtr<T>(slice.data(), slice.size());
+  }
+
+  /// from<Rust>(rustString) - Zero-copy string chars (not null-terminated)
+  static kj::ArrayPtr<const char> into(const ::rust::String& str) {
+    return kj::ArrayPtr<const char>(str.data(), str.size());
+  }
+
+  /// from<Rust>(rustStr) - Zero-copy string slice chars (not null-terminated)
+  static kj::ArrayPtr<const char> into(const ::rust::Str& str) {
+    return kj::ArrayPtr<const char>(str.data(), str.size());
+  }
+};
+
+/// Owned Rust copies: kjObject.as<RustCopy>() and from<RustCopy>(kjObject)
+struct RustCopy {
+  /// kjStringPtr.as<RustCopy>() - via RustCopy::from(&kjStringPtr)
+  static ::rust::String from(const kj::StringPtr* str) {
+    return ::rust::String(str->begin(), str->size());
+  }
+
+  /// kjArrayPtr.as<RustCopy>() - via RustCopy::from(&kjArrayPtr)
+  template <typename T>
+  static ::rust::Vec<T> from(kj::ArrayPtr<const T>* arr) {
+    ::rust::Vec<T> result;
+    result.reserve(arr->size());
+    for (auto& t: *arr) {
+      result.push_back(t);
+    }
+    return result;
+  }
+
+  /// from<RustCopy>(rustSliceOfStrs) - Copy slice of strs to null-terminated KJ strings
+  static kj::Array<kj::String> into(::rust::Slice<::rust::str> slice) {
+    auto res = kj::heapArrayBuilder<kj::String>(slice.size());
+    for (auto& entry: slice) {
+      res.add(kj::str(entry));
+    }
+    return res.finish();
+  }
+
+  /// from<RustCopy>(rustVecOfStrings) - Copy string vector to null-terminated KJ strings
+  static kj::Array<kj::String> into(const ::rust::Vec<::rust::String>& vec) {
+    auto res = kj::heapArrayBuilder<kj::String>(vec.size());
+    for (auto& entry: vec) {
+      res.add(kj::str(entry));
+    }
+    return res.finish();
+  }
+};
+
+/// Mutable Rust views: kjObject.as<RustMutable>() and from<RustMutable>(kjObject)
```

Comment:
I worry a little bit about a exposing mutable view of data to Rust while retaining mutable access from C++.

I guess it's no worse than passing a mutable reference to a C++ object to Rust, though? For there to be a problem, C++ code would have to concurrently mutate the data, which would probably look fishy on the C++ side.